designs/vpce_api_flexibility.md

```diff
@@ -0,0 +1,337 @@
+	Region string `json:"region,omitempty"`
+
+	// +kubebuilder:default=false
+	// EnablePrivateDns will allow AVO to create VPC Endpoints with private DNS names specified by a VPC Endpoint Service
+	// https://docs.aws.amazon.com/vpc/latest/privatelink/manage-dns-names.html (defaults to false)
+	EnablePrivateDns bool `json:"enablePrivateDns"`
+
+	// Unchanged
+	// ServiceName is the name of the VPC Endpoint Service to connect to
+	ServiceName string `json:"serviceName"`
+
+	// Unchanged
+	// SecurityGroup contains the configuration of the security group attached to the VPC Endpoint
+	SecurityGroup SecurityGroup `json:"securityGroup"`
+
+	// Vpc will allow AVO to use a specific VPC or use the same VPC as the ROSA cluster it's running on
+	Vpc Vpc `json:"vpc"`
+
+	// CustomDns will define configurations for all other custom DNS setups, such as a separate Route 53 Private Hosted
+	// Zone or an `ExternalName` Kubernetes service.
+	CustomDns CustomDns `json:"customDns,omitempty"`
+}
+```
+
+### SecurityGroup
+
+This is unchanged from the current API.
+
+### VPC
+
+By introducing a VPC struct, we now have the ability to specify a custom list of `SubnetIds` or continue to use the
+current implicit behavior by setting `UseRosaPrivateSubnets` to true or keeping the default value. In either case, the
+VPC ID is still inferred from the provided subnets.
+
+```go
+type Vpc struct {
+	// +kubebuilder:default=true
+	UseRosaPrivateSubnets bool     `json:"useRosaPrivateSubnets"`
+	SubnetIds             []string `json:"subnetIds,omitempty"`
+}
+```
+
+### CustomDns
+
+This is the biggest part of the refactor. Currently, the only known custom DNS configuration hinges on a
+`Route53PrivateHostedZone`. If more arise in the future, they can be added as siblings in `CustomDns`.
+
+- Like `Vpc`, `Route53PrivateHostedZone` has a boolean field `useRosaPrivateHostedZone` to maintain its existing behavior.
+- `DomainName` provides a way for consumers to specify the domain name of a new Route 53 Private Hosted Zone.
+- `Id` provides a way for consumers to specify the ID of an existing Route 53 Private Hosted Zone
+- `Record` allows for customization of the Hosted Zone Record contained in the Route 53 Private Hosted Zone
+  - `.spec.customDns.route53PrivateHostedZone.record.hostname` replaces `.spec.subdomainName` to use the same naming scheme that AWS uses.
+  - `.spec.customDns.route53PrivateHostedZone.record.externalNameService` replaces `.spec.externalNameService` as an ExternalName service is only utilized as a convenience layer for interacting with the Route 53 Private Hosted Zone Record.
+
+```go
+type CustomDns struct {
+	Route53PrivateHostedZone Route53PrivateHostedZone `json:"route53PrivateHostedZone,omitempty"`
+}
+
+type Route53PrivateHostedZone struct {
+	// +kubebuilder:default=true
+	UseRosaPrivateHostedZone bool                    `json:"useRosaPrivateHostedZone"`
+	DomainName               string                  `json:"domainName,omitempty"`
```

Comment:
Should this be ZoneDomain?

Reply:
Ah, I like `DomainName` because AWS calls this a Hosted Zone's domain name. I don't think we need to have `Zone` in the name since it'll be nested like: `.spec.customDns.route53PrivatehostedZone.domainName`